jina/types/document/graph.py

```diff
@@ -0,0 +1,322 @@
+        """
+        The in degree of the doc node
+
+        .. # noqa: DAR201
+        :param doc: the document node from which to extract the indegree.
+        """
+        in_edges = self.get_incoming_nodes(doc)
+        return len(in_edges) if in_edges else 0
+
+    @property
+    def num_nodes(self) -> int:
+        """
+        The number of nodes in the graph
+
+        .. # noqa: DAR201
+        """
+        return len(self.nodes)
+
+    @property
+    def num_edges(self) -> int:
+        """
+        The number of edges in the graph
+
+        .. # noqa: DAR201
+        """
+        adjacency = self.adjacency
+        return adjacency.data.shape[0] if adjacency is not None else 0
+
+    @property
+    def nodes(self):
+        """
+        The nodes list for this graph
+
+        .. # noqa: DAR201
+        """
+        return self.chunks
+
+    @nodes.setter
+    def nodes(self, value: Iterable['Document']):
+        """Set all nodes of the current document.
+
+        :param value: the array of nodes of this document
+        """
+        self.chunks = value
+
+    @adjacency.setter
+    def adjacency(self, value: 'coo_matrix'):
+        """
+        Set the adjacency list of this graph.
+
+        :param value: the float weight of the document.
+        """
+        SparseNdArray(self._pb_body.graph_info.adjacency, sp_format='coo').value = value
+
+    def get_outgoing_nodes(self, doc: 'Document') -> Optional[ChunkArray]:
+        """
+        Get all the outgoing edges from `doc`
+
+        .. # noqa: DAR201
+        :param doc: the document node from which to extract the outgoing nodes.
+        """
+        if self.adjacency is not None and doc.id in self._node_id_to_offset:
+            offset = self._node_id_to_offset[doc.id]
+            return ChunkArray(
+                [
+                    self.nodes[col.item()]
+                    for (row, col) in zip(self.adjacency.row, self.adjacency.col)
+                    if row.item() == offset
+                ],
+                reference_doc=self,
+            )
+
+    def get_incoming_nodes(self, doc: 'Document') -> Optional[ChunkArray]:
+        """
+        Get all the outgoing edges from `doc`
+
+        .. # noqa: DAR201
+        :param doc: the document node from which to extract the incoming nodes.
+        """
+        if self.adjacency is not None and doc.id in self._node_id_to_offset:
+            offset = self._node_id_to_offset[doc.id]
+            return ChunkArray(
+                [
+                    self.nodes[row.item()]
+                    for (row, col) in zip(self.adjacency.row, self.adjacency.col)
+                    if col.item() == offset
+                ],
+                reference_doc=self,
+            )
+
+    def __iter__(self) -> Iterator[Tuple['Document']]:
+        for (row, col) in zip(self.adjacency.row, self.adjacency.col):
+            yield self.nodes[row.item()], self.nodes[col.item()]
+
+    def __len__(self) -> int:
```

Comment:
to check if it makes sense

Reply:
Removed (and tests updated to respect this)